Add format_data_size function

nineinchnick opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

Add a `format_data_size` function, similar to `format_number` but using SI unit prefixes and bytes (`kB`, `MB` etc).

This is based on #1905

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?
new feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
core functions

> How would you describe this change to a non-technical end user or system administrator?
`format_data_size` and `format_data_size_decimal` allows formatting huge number of bytes to a human-readable format

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# General
* Add `format_data_size`, `parse_data_size_decimal`, and `format_data_size_decimal` functions. ({issue}`issuenumber`)
```

nineinchnick commented on docs/src/main/sphinx/functions/conversion.rst:
```diff
@@ -78,6 +78,13 @@ Unit    Description   Value
 ``YB``  Yottabytes    1024\ :sup:`8`
```
I think it's wrong to use 1024 in `parse_data_size`. It should either use 1000, or units should be `KiB`, `MiB` etc.

See the references on this page: https://en.wikipedia.org/wiki/Kilobyte

The question is - should I fix it in this PR or use the same invalid base in `format_data_size`? Note that changing the base in `parse_data_size` is a breaking change.

nineinchnick replied:
I added correct units in another commit.

nineinchnick replied:
The same issue exists in https://github.com/airlift/units/blob/master/src/main/java/io/airlift/units/DataSize.java but since it's used for configuration it's a really risky change to make there. Related docs page: https://trino.io/docs/current/admin/properties.html#data-size

nineinchnick replied:
I opened https://github.com/airlift/units/issues/26

dain reviewed (dismissed):
This function was added to process airlift units specifically.  The change here would break backwards compatibility of this function, and would not work for processing Trino logs anymore, so I don't think we should apply this patch to modify the existing function.  Instead, if we need this, I suggest we introduce a new function.  Either way, I would suggest discussing this on slack.

ebyhr commented:
Related Slack thread: https://trinodb.slack.com/archives/C0305TQ05KL/p1645531102538189

mosabua commented:
Whatever you do in code and function naming .. also add to https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/functions/list-by-topic.rst

And to recap from slack .. I agree that the function should be consistent with measures in other functions first .. and then if we need SI equivalents .. add more functions like format_data_size_si

nineinchnick commented:
@dain thanks, agreed about compatibility. This was the guidance I was looking for.

So I suggest the following:
* Add `parse_data_size_decimal` that would be using the correct base; I don't think we should modify `parse_data_size` in any way, like adding support for IEC units, because it would only add confusion
* Add `format_data_size` that would be using the same (invalid) base as `parse_data_size`
* Add `format_data_size_decimal` that's using the correct base and units

@mosabua thanks, I totally missed that other file. I updated it and added missing functions.

PS. All units here are SI, the distinction is between decimal SI (kilo, mega, etc.) and IEC8000 SI (kibi, mebi, etc.)

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
+        PETABYTE(new BigDecimal(1_000_000_000_000_000L)),
+        EXABYTE(new BigDecimal(1_000_000_000_000_000_000L)),
+        ZETTABYTE(new BigDecimal(1_000_000_000_000_000_000L).multiply(new BigDecimal(1_000L))),
+        YOTTABYTE(new BigDecimal(1_000_000_000_000_000_000L).multiply(new BigDecimal(1_000_000L)));
```
`new BigDecimal("1" + "0".repeat(24))` could be more readable?
the enum constructor could take the exponent only (24), so we do this once.

nineinchnick replied:
I added a `base` argument to the Unit constructor and merged decimals into it. The overall class looks better but PTAL at the constructor.

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
+        return utf8Slice(formatLong(value, 1000));
+    }
+
+    private String formatLong(double fractional, long base)
```
"fractional"? "value"?

(it's not eg "a fractional part of some value", right?)

nineinchnick replied:
Right, done!

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
+
+    private DecimalFormat getFormat(double value)
+    {
+        if (value < 10) {
+            // show up to two decimals to get 3 significant digits
+            return format3Number;
```
This should account for negative values (`values > -10 && value < 10`)

nineinchnick replied:
Added `Math.abs` around `value` and updated tests.

findepi commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
+    @Description("Formats data size as bytes using a unit symbol and a decimal base")
+    @ScalarFunction(value = "format_data_size_decimal")
+    @SqlType(StandardTypes.VARCHAR)
+    public Slice formatDataSizeDecimal(@SqlType(StandardTypes.DOUBLE) double value)
```
What does this method return when invoked with +/- infinity?
What does it return when invoked with NaN?

nineinchnick replied:
Thanks, I missed this. I added conditions to avoid formatting these values (avoid adding a suffix, since `InfinityB` or `NaNB` doesn't make sense) and added test cases.

nineinchnick commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -21,18 +21,34 @@
+
+    public DataSizeFunctions()
+    {
+        format3Number = new DecimalFormat("#.##");
```
I copied this from the `format_number` function and it was recommended there by David in https://github.com/trinodb/trino/pull/1905#discussion_r341802228

nineinchnick commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
+    }
+
+    @Description("Formats data size as bytes using a unit symbol")
+    @ScalarFunction(value = "format_data_size")
```
Done!

nineinchnick commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -133,4 +172,140 @@ public static Unit parse(String unitString, String dataSize)
+    @SqlType(StandardTypes.VARCHAR)
+    public Slice formatDataSizeDecimal(@SqlType(StandardTypes.BIGINT) long value)
+    {
+        return utf8Slice(formatLong((double) value, 1000));
```
Done!

dain commented:
This needs syntax review for the new functions.

nineinchnick commented:
Who's doing the syntax review? Can I do anything to help with this kind of review?

bitsondatadev commented:
:wave: @nineinchnick - this PR has become inactive. If you're still interested in working on it, please let us know.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

nineinchnick commented:
There's nothing for me to do here, I'm waiting for syntax review.

bitsondatadev commented:
Okay I'll see if @dain can take a look!

martint commented on core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java:
```diff
@@ -41,7 +57,39 @@ private DataSizeFunctions() {}
+    @ScalarFunction("parse_data_size_decimal")
+    @LiteralParameters("x")
+    @SqlType("decimal(38,0)")
+    public static Int128 parseDataSizeDecimal(@SqlType("varchar(x)") Slice input)
```
What's the difference between this function and the one above?

nineinchnick replied:
This one is using base 1000, the other one base 1024. Both are using SI units (kilo, mega), not IEC8000 units (kibi, mebi), to avoid any backward compatibility issues.

nineinchnick commented:
@martint PTAL. If there's no interest in merging this, I'll close it in 2 weeks.

nineinchnick commented:
Since we have SQL routines now, it's trivial to implement formatting functions using them.

mosabua commented:
Hahah ... well.. I got something for you then .. just wait for the next Trino blog post coming tomorrow.

nineinchnick commented:
```sql
-- format_data_size is like format_number but uses SI units (base 1024)
CREATE OR REPLACE FUNCTION format_data_size(input BIGINT)
    RETURNS VARCHAR
    DETERMINISTIC
    RETURNS NULL ON NULL INPUT
    BEGIN
        DECLARE value DOUBLE DEFAULT CAST(input AS DOUBLE);
        DECLARE result BIGINT;
        DECLARE base INT DEFAULT 1024;
        DECLARE unit VARCHAR DEFAULT 'B';
        DECLARE format VARCHAR;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'kB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'MB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'GB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'TB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'PB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'EB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'ZB';
        END IF;
        IF abs(value) >= base THEN
            SET value = value / base;
            SET unit = 'YB';
        END IF;
        IF abs(value) < 10 THEN
            SET format = '%.2f';
        ELSEIF abs(value) < 100 THEN
            SET format = '%.1f';
        ELSE
            SET format = '%.0f';
        END IF;
        RETURN format(format, value) || unit;
    END;
```
